core/iwasm/interpreter/wasm_runtime.c

```diff
@@ -2268,6 +2268,37 @@ wasm_instantiate(WASMModule *module, WASMModuleInstance *parent,
     }
 #endif
 
+#if WASM_ENABLE_SHARED_HEAP != 0
+    if (!is_sub_inst) {
+        uint32 shared_heap_size = wasm_runtime_get_shared_heap_size();
+        uint32 heap_struct_size = mem_allocator_get_heap_struct_size();
+
+        if (!(module_inst->shared_heap = runtime_malloc(
+                  (uint64)sizeof(WASMSharedHeap), error_buf, error_buf_size))) {
+            goto fail;
+        }
+
+        if (!(module_inst->shared_heap->handle = runtime_malloc(
+                  (uint64)heap_struct_size, error_buf, error_buf_size))) {
+            goto fail;
+        }
+
+        if (!(module_inst->shared_heap->data = runtime_malloc(
+                  (uint64)shared_heap_size, error_buf, error_buf_size))) {
+            goto fail;
+        }
+
+        module_inst->shared_heap->size = shared_heap_size;
+
+        if (!mem_allocator_create_with_struct_and_pool(
```

Comment:
doesn't this allocator put metadata on the heap itself?
if so, i guess it isn't appropriate for shared heap.

Reply:
The heap is inited by module_inst->shared_heap->data, and in future implementation, module_inst->shared_heap->data is to be shared between multi modules

Reply:
i meant that, if metadata is stored on the shared heap itself, any modules accessible to the heap can easily corrupt the metadata.
is it ok for your use cases?

Reply:
Because mem_allocator_create_with_struct_and_pool rather than mem_allocator_create is used to create a allocator, the metadata of allocator is separated from the heap